src/Agent.Sdk/LogPlugin.cs

```diff
@@ -0,0 +1,534 @@
+                    {
+                        _trace.Trace($"Plugin '{queue.Key}' has too many buffered outputs.");
+                        flag[queue.Key]++;
+                        if (flag[queue.Key] >= 10)
+                        {
+                            _trace.Trace($"Short circuit plugin '{queue.Key}'.");
+                            _shortCircuited[queue.Key].TrySetResult(0);
+                        }
+                    }
+                    else if (flag[queue.Key] > 0)
+                    {
+                        _trace.Trace($"Plugin '{queue.Key}' has cleared out buffered outputs.");
+                        flag[queue.Key] = 0;
+                    }
+                }
+
+                await Task.WhenAny(Task.Delay(_shortCircuitMonitorFrequency), Task.Delay(-1, token));
+            }
+        }
+
+        private async Task RunAsync(IAgentLogPlugin plugin, CancellationToken token)
+        {
+            List<string> errors = new List<string>();
+            string typeName = plugin.GetType().FullName;
+            var context = _pluginContexts[typeName];
+
+            bool initialized = false;
+            try
+            {
+                initialized = await plugin.InitializeAsync(context);
+            }
+            catch (Exception ex)
+            {
+                errors.Add($"Fail to initialize: {ex.Message}.");
+                context.Trace(ex.ToString());
+            }
+            finally
+            {
+                if (!initialized)
+                {
+                    context.Trace("Skip process outputs base on plugin initialize result.");
+                    _shortCircuited[typeName].TrySetResult(0);
+                }
+            }
+
+            using (var registration = token.Register(() =>
+                                      {
+                                          var depth = _outputQueue[typeName].Count;
+                                          if (depth > 0)
+                                          {
+                                              context.Output($"Waiting for log plugin to finish, pending process {depth} log lines.");
+                                          }
+                                      }))
+            {
+                while (!_shortCircuited[typeName].Task.IsCompleted && !token.IsCancellationRequested)
+                {
+                    while (!_shortCircuited[typeName].Task.IsCompleted && _outputQueue[typeName].TryDequeue(out string line))
+                    {
+                        try
+                        {
+                            var id = line.Substring(0, line.IndexOf(":"));
+                            var message = line.Substring(line.IndexOf(":") + 1);
+                            var processLineTask = plugin.ProcessLineAsync(context, _steps[id], message);
+                            var completedTask = await Task.WhenAny(_shortCircuited[typeName].Task, processLineTask);
```

Comment:
I don't think I understand the purpose of this code. 😄

`Task.WhenAny()` returns the task that completed; `await`-ing such task afterwards would basically be a no-op. What's the goal of the code here?

Reply:
if there is a exception, you need to `await` the completed task to let it throw.